File: core/src/rules/validate-type.ts
// core/src/rules/validate-type.ts
//
// Atomic rule: validates a JSON-LD object's @type.
//
// Outcomes:
//   - missing (undefined / null / empty)     → MISSING_TYPE  (error)
//   - non-string (number, array, object, …)  → UNKNOWN_TYPE  (error)
//   - string not in registry                 → UNKNOWN_TYPE  (error)
//   - string in registry                     → no issue
//
// Pure function: no I/O, no shared state, no side effects.
//
// Cycle 1 supports @type as a single string only. Multi-typed entities
// (@type: ["Article", "BlogPosting"]) are deferred to a later cycle.

import type { Issue } from "../types.js";
import type { Registry } from "../registry.js";
import { missingType, unknownType } from "../errors.js";

export function validateType(value: unknown, registry: Registry): Issue[] {
  if (value === undefined || value === null || value === "") {
    return [missingType()];
  }

  if (typeof value !== "string") {
    return [unknownType(value)];
  }

  if (registry.types[value] === undefined) {
    return [unknownType(value)];
  }

  return [];
}

File: python/src/schema_audit/rules/validate_type.py
"""Atomic rule: validate a JSON-LD ``@type`` value.

Mirrors ``core/src/rules/validate-type.ts`` 1:1.

Outcomes:

- missing (None / empty string)            → ``MISSING_TYPE`` (error)
- non-string (number, list, dict, …)       → ``UNKNOWN_TYPE`` (error)
- string not in registry                   → ``UNKNOWN_TYPE`` (error)
- string in registry                       → no issue

Cycle 9 supports ``@type`` as a single string only. Multi-typed
entities (``@type: ["Article", "BlogPosting"]``) are handled by the
orchestrator before this rule sees the value.
"""

from __future__ import annotations

from ..errors import missing_type, unknown_type
from ..registry import Registry
from ..types import Issue


def validate_type(value: object, registry: Registry) -> list[Issue]:
    if value is None or value == "":
        return [missing_type()]

    if not isinstance(value, str):
        return [unknown_type(value)]

    if value not in registry["types"]:
        return [unknown_type(value)]

    return []
